### PR TITLE
Refactoring

### PR DESCRIPTION
**Pull request by Hyperkid123**
jira: https://issues.redhat.com/browse/RHCLOUD-21058

### Changes
- allow multiple templates per service
- move all source files to /src folder
- fix tsconfig config
- remove unnecessary webpack settings
- migrate all files to TS and forbid JS files
- use jest-puppeteer extension
- configure and fix all eslint warnings
- add a circular dependency check
- add context middleware to get the user identity
- move env variables to global config for one source of truth

### CODE REVIEW

File: src/server/routes/routes.ts [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+import ServiceNames from '../../common/service-names';
+import renderTemplate from '../render-template';
+import { processOrientationOption } from '../../browser/helpers';
+import generatePdf, { previewPdf } from '../../browser';
+import { SendingFailedError, PDFNotFoundError } from '../errors';
+import config from '../../common/config';
+import { OPTIONS_HEADER_NAME } from '../../common/consts';
+import { PreviewReqQuery } from '../../common/types';
+
+const PORT = config.webPort;
```
Comment (karelhala) [inline src/server/routes/routes.ts:15]: Wouldn't it be better to consume these from env variables?

File: src/server/routes/routes.ts [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+    unknown,
+    unknown,
+    { service: ServiceNames; template: string }
+  > {
+  headers: IncomingHttpHeaders & {
+    [OPTIONS_HEADER_NAME]?: string;
+  };
+}
+
+const APIPrefix = '/api/crc-pdf-generator/v1';
```
Comment (karelhala) [inline src/server/routes/routes.ts:51]: How about populating this from env variable?

File: src/server/routes/routes.ts [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+});
+
+router.get(`${APIPrefix}/hello`, (_req, res) => {
+  return res.status(200).send('<h1>Well this works!</h1>');
+});
+
+router.post(
+  `${APIPrefix}/generate`,
+  async (req: GenerateHandlerReqyest, res) => {
+    const rhIdentity = req.headers['x-rh-identity'] as string;
```
Comment (karelhala) [inline src/server/routes/routes.ts:102]: Aren't there some kind of interceptors to bounce us away on any request? So we don't have to remeber to repeat this all the time.

File: src/server/routes/routes.ts [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+      return res.status(401).send('Unauthorized access not allowed');
+    }
+
+    const service = req.body.service;
+    const template = req.body.template;
+    const dataOptions = req.body;
+
+    const tenant = JSON.parse(atob(rhIdentity))['identity']['internal'][
+      'org_id'
+    ];
```
Comment (karelhala) [inline src/server/routes/routes.ts:115]: Same for this, isn't there a dependency which would parse this and passed it to each method?

File: src/templates/compliance/components/SystemsTableHeader.tsx [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+  nonCompliantSystemCount: number;
+  unsupportedSystemCount: number;
+}
+
+const SystemsTableHeader: React.FC<HeaderTable> = ({
+  compliantSystemCount,
+  nonCompliantSystemCount,
+  unsupportedSystemCount,
+}) => {
+  return (
```
Comment (karelhala) [inline src/templates/compliance/components/SystemsTableHeader.tsx:22]: Extra return statement

**Review by karelhala — COMMENTED** [review]
Looking good! Just a few minor questions.

File: src/server/index.ts [hunk 6]
```diff
@@ -0,0 +1,53 @@
+import express from 'express';
```
[the remark this answers is not in the document]
Comment (Hyperkid123) [reply src/server/index.ts:1]: Interesting idea. Although I think 3 scale should handle most of the potential vulnerabilities.

File: src/server/routes/routes.ts (hunk 3 above)
Comment (Hyperkid123) [reply src/server/routes/routes.ts:102]: We can use middleware and set it in context https://www.npmjs.com/package/express-http-context

File: src/server/routes/routes.ts (hunk 2 above)
Comment (Hyperkid123) [reply src/server/routes/routes.ts:51]: Again I would probably move this to the config variable as ENV variables are not really a thing. Maybe I should do the same for the prod/dev flags.

File: src/server/routes/routes.ts (hunk 1 above)
Comment (Hyperkid123) [reply src/server/routes/routes.ts:15]: In prod, these variables are set by clowder at runtime :( So no env variables I am afraid.

File: src/server/routes/routes.ts (hunk 4 above)
Comment (Hyperkid123) [reply src/server/routes/routes.ts:115]: This is actually useless value. Does nothing just logs the header. And we can use header for that.

File: src/server/routes/routes.ts (hunk 3 above)
Comment (karelhala) [reply src/server/routes/routes.ts]: Please do! I am all in favor of simplifiying these requests.

**Comment by karelhala** [comment]
@Hyperkid123 how about openapi docs? Do we want them to add them now, or later? We can by all means create new ticket just for that.

File: src/server/routes/routes.ts (hunk 1 above)
Comment (Hyperkid123) [reply src/server/routes/routes.ts]: It takes the variables from a file and propagates them through a dependency ('app-common-js'). We could then slap the vars into env, but the issue is that some variables (like other services endpoints) are not strings or different primitive types. So we would have to parse it anyways. And I'd like to do it once after the service startup, rather then every time we want to use a variable.

**Comment by Hyperkid123** [comment]
> @Hyperkid123 how about openapi docs? Do we want them to add them now, or later? We can by all means create new ticket just for that.

Later on, I just wanted to get the codebase up to date.

File: src/server/routes/routes.ts (hunk 3 above)
Comment (Hyperkid123) [reply src/server/routes/routes.ts]: Its done

**Pull request merged**